transport/http_transport.go

```diff
@@ -0,0 +1,467 @@
+	return t.listener.Addr().String()
+}
+
+// ServeHTTP handles all incoming HTTP requests received by the server spawned by SpawnServer()
+func (t *HttpTransport) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if t.instance == nil {
+		panic("groupcache instance is nil; you must register an instance by calling HttpTransport.Register()")
+	}
+
+	if !strings.HasPrefix(r.URL.Path, t.opts.BasePath) {
+		panic("HTTPPool serving unexpected path: " + r.URL.Path)
+	}
+
+	parts := strings.SplitN(r.URL.Path[len(t.opts.BasePath):], "/", 2)
+	if len(parts) != 2 {
+		http.Error(w, "bad request", http.StatusBadRequest)
+		return
+	}
+	groupName := parts[0]
+	key := parts[1]
+
+	type transportMethods interface {
+		Get(ctx context.Context, key string, dest Sink) error
+		LocalSet(string, []byte, time.Time)
+		LocalRemove(string)
+	}
+
+	// Fetch the value for this group/key.
+	group := t.instance.GetGroup(groupName).(transportMethods)
+	if group == nil {
+		http.Error(w, "no such group: "+groupName, http.StatusNotFound)
+		return
+	}
+	var ctx context.Context
+	if t.opts.Context != nil {
+		ctx = t.opts.Context(r)
+	} else {
+		ctx = r.Context()
+	}
+
+	// Delete the key and return 200
+	if r.Method == http.MethodDelete {
+		group.LocalRemove(key)
+		return
+	}
+
+	// The read the body and set the key value
+	if r.Method == http.MethodPut {
+		defer r.Body.Close()
+		b := bufferPool.Get().(*bytes.Buffer)
+		b.Reset()
+		defer bufferPool.Put(b)
+		_, err := io.Copy(b, r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		var out pb.SetRequest
+		err = proto.Unmarshal(b.Bytes(), &out)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		var expire time.Time
+		if out.Expire != nil && *out.Expire != 0 {
+			expire = time.Unix(*out.Expire/int64(time.Second), *out.Expire%int64(time.Second))
+		}
+
+		group.LocalSet(*out.Key, out.Value, expire)
+		return
+	}
+
+	var b []byte
+
+	value := AllocatingByteSliceSink(&b)
+	err := group.Get(ctx, key, value)
+	if err != nil {
+		if errors.Is(err, &ErrNotFound{}) {
+			http.Error(w, err.Error(), http.StatusNotFound)
+			return
+		}
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+
+	view, err := value.View()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	var expireNano int64
+	if !view.Expire().IsZero() {
+		expireNano = view.Expire().UnixNano()
+	}
+
+	// Write the value to the response body as a proto message.
+	body, err := proto.Marshal(&pb.GetResponse{Value: b, Expire: &expireNano})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", "application/x-protobuf")
+	_, _ = w.Write(body)
+}
+
+// NewClient creates a new http client for the provided peer
+func (t *HttpTransport) NewClient(_ context.Context, p peer.Info) (peer.Client, error) {
```

Comment:
Just to ask: is there a common transport where a context is needed for client creation?

Reply:
This is here to support third party transports that are not a part of `groupcache` that might need to "connect" or preform some operation that may need to be timed out by the caller.

Reply:
Got ita